### PR TITLE
DOMA-2173 employee division limited visibility

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**
Backend for [task](https://doma.atlassian.net/browse/DOMA-2173)

Limited the visibility of `tickets`, `meters` and `meter readings` for employees who has roles that have `canReadEntitiesOnlyInScopeOfDivision` is true.

### CODE REVIEW

File: apps/condo/domains/organization/schema/OrganizationEmployeeRole.js [hunk 1]
```diff
@@ -74,6 +74,11 @@ const OrganizationEmployeeRole = new GQLListSchema('OrganizationEmployeeRole', {
             type: Checkbox,
             defaultValue: true,
         },
+        isDivisionLimitedVisibility: {
```
Comment (AntonAL) [inline apps/condo/domains/organization/schema/OrganizationEmployeeRole.js:77]: Maybe, this name will better reflect the logic behind this field?
Is this limitation only for reading operation?
```suggestion
        canReadEntitiesOnlyInScopeOfDivision: {
```

Will this limitation cover update, delete and create operations?
For this case, my suggestion is following:
```suggestion
        canOperateWithEntitiesOnlyInScopeOfDivision: {
```

File: apps/condo/domains/division/schema/Division.js [hunk 2]
```diff
@@ -72,6 +73,10 @@ const Division = new GQLListSchema('Division', {
                     const [organization] = await Organization.getAll(context, { id: organizationId })
 
                     const propertyIds = resolvedData[fieldPath]
+                    if (propertyIds.length > MAX_PROPERTIES_IN_DIVISION) {
+                        addFieldValidationError('Exceeded the maximum number of properties in the division (150 properties)')
```
Comment (pahaz) [inline apps/condo/domains/division/schema/Division.js:77]: @AntonAL should we use your new error style here?

File: apps/condo/domains/division/constants.js [hunk 3]
```diff
@@ -2,5 +2,8 @@
  * Generated by `createschema division.Division 'name:Text; organization:Relationship:Organization:CASCADE; responsible:Relationship:OrganizationEmployee:PROTECT;'`
  */
 
+const MAX_PROPERTIES_IN_DIVISION = 150
```
[the remark this answers is not in the document]
Comment (pahaz) [reply apps/condo/domains/division/constants.js:5]: It's a business-level rule. in a good way, we should have such restrictions for all fields

File: apps/condo/domains/division/utils/serverSchema/index.js [hunk 4]
```diff
@@ -7,12 +7,40 @@
...
 const { Division: DivisionGQL } = require('@condo/domains/division/gql')
+const { OrganizationEmployee } = require('@condo/domains/organization/utils/serverSchema')
+const uniq = require('lodash/uniq')
+const flatten = require('lodash/flatten')
 /* AUTOGENERATE MARKER <IMPORT> */
 
 const Division = generateServerUtils(DivisionGQL)
 /* AUTOGENERATE MARKER <CONST> */
 
+async function getUserDivisionsInfo (context, userId) {
```
Comment (pahaz) [inline apps/condo/domains/division/utils/serverSchema/index.js:18]: it's better to use low level getByCondition if you need to return just an IDs

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/organization/schema/OrganizationEmployeeRole.js (hunk 1 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/organization/schema/OrganizationEmployeeRole.js:77]: This limitation only for reading operation.
Agree that the name `canReadEntitiesOnlyInScopeOfDivision` is better

File: apps/condo/domains/meter/schema/Meter.test.js [hunk 5]
```diff
@@ -548,6 +550,73 @@ describe('Meter', () => {
             expect(meters).toHaveLength(0)
         })
 
+        test('employee with division limited role: can read meters only from his divisions', async () => {
```
Comment (AntonAL) [inline apps/condo/domains/meter/schema/Meter.test.js:553]: Please, introduce a `describe` blocks for employee with and without division scoped visibility.

File: apps/condo/domains/meter/schema/MeterReading.test.js [hunk 6]
```diff
@@ -185,6 +186,81 @@ describe('MeterReading', () => {
             })
         })
 
+        test('employee with division limited role: can read meters only from his divisions', async () => {
```
Comment (AntonAL) [inline apps/condo/domains/meter/schema/MeterReading.test.js:189]: Please, introduce a `describe` blocks for employee with and without division scoped visibility.

For example:

```
describe('employee with division scoped visibility', () => {
  it('can read meters only from his divisions')
})

describe('employee without division scoped visibility', () => {
  it('can read all …')
})
```

**Review by AntonAL — APPROVED** [review]

**Pull request merged**